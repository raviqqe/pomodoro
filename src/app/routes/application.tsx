import { styled } from "@linaria/react";
import { SignOut } from "../components/SignOut.js";
import { ViewGraphButton } from "../components/ViewGraphButton.js";
import { ViewTimerButton } from "../components/ViewTimerButton.js";
import { Outlet, useLocation } from "react-router";

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

const Buttons = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;

  > :not(:last-child) {
    margin-bottom: 0.5em;
  }
`;

export default (): JSX.Element => {
  const { pathname } = useLocation();

  return (
    <Container>
      <Outlet />
      <Buttons>
        <SignOut />
        {pathname === "/timer" ? <ViewGraphButton /> : <ViewTimerButton />}
      </Buttons>
    </Container>
  );
};
