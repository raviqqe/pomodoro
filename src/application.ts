export type * from "./application/performance-graph.js";
