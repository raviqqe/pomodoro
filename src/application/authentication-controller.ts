export interface IAuthenticationController {
  signIn(): Promise<boolean>;
  signOut(): Promise<boolean>;
  isSignedIn(): Promise<boolean>;
}
