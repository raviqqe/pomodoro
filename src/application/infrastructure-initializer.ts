export interface IInfrastructureInitializer {
  initialize(): Promise<void>;
}
