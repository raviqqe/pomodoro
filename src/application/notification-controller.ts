export interface NotificationInitializer {
  initialize(): Promise<void>;
}
