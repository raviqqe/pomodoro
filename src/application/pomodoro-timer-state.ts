export enum PomodoroTimerState {
  Pomodoro,
  ShortBreak,
  LongBreak,
}
