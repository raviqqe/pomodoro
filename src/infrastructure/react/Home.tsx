import React from "react";
import { PulseLoader } from "react-spinners";
import styled from "styled-components";
import { SignOut } from "./SignOut";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  width: 85ex;
  max-width: 100%;
  margin: auto;

  > :first-child {
    flex: 1;
  }
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const SignOutContainer = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;
`;

export interface IProps {
  signOut: () => Promise<void>;
}

export const Home = ({ signOut }: IProps) => {
  return (
    <Container>
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
      <SignOutContainer>
        <SignOut signOut={signOut} />
      </SignOutContainer>
    </Container>
  );
};
