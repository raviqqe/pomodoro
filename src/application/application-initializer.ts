import { IAuthenticationController } from "./authentication-controller";
import { IInfrastructureInitializer } from "./infrastructure-initializer";

export interface IInitialState {
  signedIn: boolean;
}

export class ApplicationInitializer {
  constructor(
    private readonly infrastructureInitializer: IInfrastructureInitializer,
    private readonly authenticationController: IAuthenticationController
  ) {}

  public async initialize(): Promise<IInitialState> {
    await this.infrastructureInitializer.initialize();
    return { signedIn: await this.authenticationController.isSignedIn() };
  }
}
