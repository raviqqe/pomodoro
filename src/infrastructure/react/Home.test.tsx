import {
  render,
  fireEvent,
  waitFor,
  RenderResult,
  act,
} from "@testing-library/react";
import { create } from "react-test-renderer";
import { PomodoroTimerState } from "../../../application/pomodoro-timer-state";
import { Home, IProps } from "../Home";

const props: IProps = {
  performanceGraph: { data: [] },
  signOut: async () => {},
  startTimer: async () => {},
  stopTimer: async () => {},
  timer: { seconds: 42, state: PomodoroTimerState.Pomodoro, stopped: false },
  viewGraph: async () => {},
};

it("renders", () => {
  expect(create(<Home {...props} />).toJSON()).toMatchSnapshot();
});

it("views a performance graph", async () => {
  let result: RenderResult | undefined;

  act(() => {
    result = render(<Home {...props} />);
  });

  act(() => {
    fireEvent.click(
      result?.container.querySelector(`[aria-label="View Graph"]`) as Element
    );
  });

  await waitFor(() =>
    expect(
      result?.container.querySelector(`[aria-label="View Timer"]`)
    ).toBeTruthy()
  );

  expect(result?.container).toMatchSnapshot();
});

it("goes back to a timer view", async () => {
  let result: RenderResult | undefined;

  act(() => {
    result = render(<Home {...props} />);
  });

  act(() => {
    fireEvent.click(
      result?.container.querySelector(`[aria-label="View Graph"]`) as Element
    );
  });

  await waitFor(() =>
    expect(
      result?.container.querySelector(`[aria-label="View Timer"]`)
    ).toBeTruthy()
  );

  act(() => {
    fireEvent.click(
      result?.container.querySelector(`[aria-label="View Timer"]`) as Element
    );
  });

  await waitFor(() =>
    expect(
      result?.container.querySelector(`[aria-label="View Graph"]`)
    ).toBeTruthy()
  );

  expect(result?.container).toMatchSnapshot();
});
