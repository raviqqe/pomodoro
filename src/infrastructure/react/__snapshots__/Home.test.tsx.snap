// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`goes back to a timer view 1`] = `
<div>
  <div
    class="sc-jrcSnH dcXPTj"
  >
    <div
      class="sc-jSVbBN htDmhB"
    >
      <div
        class="sc-iBYQGD jnsiEs"
      >
        <span
          class="sc-ftTHFa"
        >
          0
        </span>
        <span
          class="sc-pyfiu laaNyd"
        >
          42
        </span>
      </div>
      <button
        class="sc-dkrGVk sc-eDvSBu iMrLIa enYyxP"
      >
        Stop
      </button>
    </div>
    <div
      class="sc-kDvvrc vaqM"
    >
      <button
        class="sc-dkrGVk sc-hLBbCX iMrLIa eva-DXt"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5a2 2 0 00-2 2v4h2V5h14v14H5v-4H3v4a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
      <button
        aria-label="View Graph"
        class="sc-dkrGVk sc-hLBbCX iMrLIa eva-DXt"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M176 64h64v384h-64zM80 336h64v112H80zM272 272h64v176h-64zM368 176h64v272h-64z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders 1`] = `
<div
  class="sc-jrcSnH dcXPTj"
>
  <div
    class="sc-jSVbBN htDmhB"
  >
    <div
      class="sc-iBYQGD jnsiEs"
    >
      <span
        class="sc-ftTHFa"
      >
        0
      </span>
      <span
        class="sc-pyfiu laaNyd"
      >
        42
      </span>
    </div>
    <button
      class="sc-dkrGVk sc-eDvSBu iMrLIa enYyxP"
    >
      Stop
    </button>
  </div>
  <div
    class="sc-kDvvrc vaqM"
  >
    <button
      class="sc-dkrGVk sc-hLBbCX iMrLIa eva-DXt"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5a2 2 0 00-2 2v4h2V5h14v14H5v-4H3v4a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </button>
    <button
      aria-label="View Graph"
      class="sc-dkrGVk sc-hLBbCX iMrLIa eva-DXt"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M176 64h64v384h-64zM80 336h64v112H80zM272 272h64v176h-64zM368 176h64v272h-64z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`views a performance graph 1`] = `
<div>
  <div
    class="sc-jrcSnH dcXPTj"
  >
    <div
      class="sc-gswPWN bpexcZ"
    >
      No performance graph to show yet!
    </div>
    <div
      class="sc-kDvvrc vaqM"
    >
      <button
        class="sc-dkrGVk sc-hLBbCX iMrLIa eva-DXt"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5a2 2 0 00-2 2v4h2V5h14v14H5v-4H3v4a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
      <button
        aria-label="View Timer"
        class="sc-dkrGVk sc-hLBbCX iMrLIa eva-DXt"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 1h6v2H9zM19.03 7.39l1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42A8.962 8.962 0 0012 4a9 9 0 00-9 9c0 4.97 4.02 9 9 9a8.994 8.994 0 007.03-14.61zM13 14h-2V8h2v6z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
